packages/shared/areHookInputsEqual.js

```diff
@@ -14,15 +14,17 @@ export default function areHookInputsEqual(arr1: any[], arr2: any[]) {
   // Don't bother comparing lengths in prod because these arrays should be
   // passed inline.
   if (__DEV__) {
-    warning(
-      arr1.length === arr2.length,
-      'Detected a variable number of hook dependencies. The length of the ' +
-        'dependencies array should be constant between renders.\n\n' +
-        'Previous: %s\n' +
-        'Incoming: %s',
-      arr1.join(', '),
-      arr2.join(', '),
-    );
+    if (arr1 === null || arr2 === null || arr1.length !== arr2.length) {
+      warning(
+        false,
+        'Detected a variable number of hook dependencies. The length of the ' +
+          'dependencies array should be constant between renders.\n\n' +
+          'Previous: %s\n' +
+          'Incoming: %s',
+        arr1 === null ? '(empty)' : `[${arr1.join(', ')}]`,
+        arr2 === null ? '(empty)' : `[${arr2.join(', ')}]`,
+      );
+    }
   }
   for (let i = 0; i < arr1.length; i++) {
```

Comment:
Somewhere around here I'm betting the VM will insert a type check anyway. I bet we can get away with just doing a proper null check on the one we're not sure about. That way you don't have to cheat the type system.